// @reactivex/rxjs/dist/amd/operators/repeat.js
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
define(["require", "exports", '../Subscriber', '../schedulers/immediate'], function (require, exports, Subscriber_1, immediate_1) {
    function repeat(count) {
        if (count === void 0) { count = -1; }
        return this.lift(new RepeatOperator(count, this));
    }
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.default = repeat;
    var RepeatOperator = (function () {
        function RepeatOperator(count, original) {
            this.count = count;
            this.original = original;
        }
        RepeatOperator.prototype.call = function (subscriber) {
            return new RepeatSubscriber(subscriber, this.count, this.original);
        };
        return RepeatOperator;
    })();
    var RepeatSubscriber = (function (_super) {
        __extends(RepeatSubscriber, _super);
        function RepeatSubscriber(destination, count, original) {
            _super.call(this, destination);
            this.count = count;
            this.original = original;
            this.invalidateRepeat();
        }
        RepeatSubscriber.prototype.repeatSubscription = function () {
            var state = { dest: this.destination, count: this.count, original: this.original };
            immediate_1.default.scheduleNow(RepeatSubscriber.dispatchSubscription, state);
        };
        RepeatSubscriber.prototype.invalidateRepeat = function () {
            var completed = this.count === 0;
            if (completed) {
                this.destination.complete();
            }
            return completed;
        };
        RepeatSubscriber.dispatchSubscription = function (_a) {
            var dest = _a.dest, count = _a.count, original = _a.original;
            return original.subscribe(new RepeatSubscriber(dest, count, original));
        };
        RepeatSubscriber.prototype._complete = function () {
            if (!this.invalidateRepeat()) {
                this.count--;
                this.repeatSubscription();
            }
        };
        return RepeatSubscriber;
    })(Subscriber_1.default);
});
